core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,93 +211,573 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
+    ZManaged.absolve(mapError(ev1).map(ev2(_).toRight(())))
+
+  /**
+   * Depending on the environment execute this or the other effect
+   */
+  final def join[R1, E1 >: E, A1 >: A](that: ZManaged[R1, E1, A1]): ZManaged[Either[R, R1], E1, A1] = self ||| that
+
+  final def left[R1 <: R, C]: ZManaged[Either[R1, C], E, Either[A, C]] = self +++ ZManaged.identity
+
+  /**
+   * Returns an effect whose success is mapped by the specified `f` function.
+   */
+  final def map[B](f0: A => B): ZManaged[R, E, B] =
+    ZManaged[R, E, B] {
+      reserve.map(token => token.copy(acquire = token.acquire.map(f0)))
+    }
+
+  /**
+   * Returns an effect whose failure is mapped by the specified `f` function.
+   */
+  final def mapError[E1](f: E => E1): ZManaged[R, E1, A] =
+    Managed(reserve.mapError(f).map(r => Reservation(r.acquire.mapError(f), r.release)))
+
+  /**
+   * Executes this effect, skipping the error but returning optionally the success.
+   */
+  final def option: ZManaged[R, Nothing, Option[A]] =
+    fold(_ => None, Some(_))
+
+  /**
+   * Translates effect failure into death of the fiber, making all failures unchecked and
+   * not a part of the type of the effect.
+   */
+  final def orDie(implicit ev: E <:< Throwable): ZManaged[R, Nothing, A] =
+    orDieWith(ev)
+
+  /**
+   * Keeps none of the errors, and terminates the fiber with them, using
+   * the specified function to convert the `E` into a `Throwable`.
+   */
+  final def orDieWith(f: E => Throwable): ZManaged[R, Nothing, A] =
+    mapError(f).catchAll(ZManaged.die)
+
+  /**
+   * Executes this effect and returns its value, if it succeeds, but
+   * otherwise executes the specified effect.
+   */
+  final def orElse[R1 <: R, E2, A1 >: A](that: => ZManaged[R1, E2, A1]): ZManaged[R1, E2, A1] =
+    foldM(_ => that, ZManaged.succeed)
+
+  /**
+   * Returns an effect that will produce the value of this effect, unless it
+   * fails, in which case, it will produce the value of the specified effect.
+   */
+  final def orElseEither[R1 <: R, E2, B](that: => ZManaged[R1, E2, B]): ZManaged[R1, E2, Either[A, B]] =
+    foldM(_ => that.map(Right[A, B]), a => ZManaged.succeed(Left[A, B](a)))
+
+  /**
+   * Provides the `ZManaged` effect with its required environment, which eliminates
+   * its dependency on `R`.
+   */
+  final def provide(r: R): ZManaged[Any, E, A] =
+    provideSome(_ => r)
+
+  /**
+   * Provides some of the environment required to run this effect,
+   * leaving the remainder `R0`.
+   */
+  final def provideSome[R0](f: R0 => R): ZManaged[R0, E, A] =
+    Managed(reserve.provideSome(f).map(r => Reservation(r.acquire.provideSome(f), r.release.provideSome(f))))
+
+  /**
+   * Keeps some of the errors, and terminates the fiber with the rest.
+   */
+  final def refineOrDie[E1](pf: PartialFunction[E, E1])(implicit ev: E <:< Throwable): ZManaged[R, E1, A] =
+    refineOrDie(pf)(ev)
```

Comment:
Sorry, I meant that right now that function will be calling itself ad infinitum. Indeed the [one in `ZIO`](https://github.com/scalaz/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/ZIO.scala#L613-L614) is:
```suggestion
    refineOrDieWith(pf)(ev)
```

Reply:
Oh good catch! Thank you